tlatools/org.lamport.tlatools/test/util/ParserAPI.java

```diff
@@ -0,0 +1,133 @@
+ * tests. This class can be viewed as a prototype of a future API to be
+ * integrated into SANY itself.
+ */
+public class ParserAPI {
+
+  /**
+   * Fully parses and checks the given source code input.
+   *
+   * @param The TLA+ source to parse, as a string.
+   * @return A level-checked semantic parse tree, or null if failure.
+   */
+  public static ModuleNode parse(String input) {
+    SyntaxTreeNode syntaxRoot = processSyntax(input);
+    if (null == syntaxRoot) {
+      return null;
+    }
+    Errors log = new Errors();
+    ModuleNode semanticRoot = processSemantics(syntaxRoot, log);
+    if (log.isFailure() || null == semanticRoot) {
+      ToolIO.err.println(log.toString());
+      return null;
+    }
+    boolean levelOk = checkLevel(semanticRoot, log);
+    if (!levelOk || log.isFailure()) {
+      ToolIO.err.println(log.toString());
+      return null;
+    }
+    return semanticRoot;
+  }
+
+  /**
+   * Checks the syntax of the given input and builds a parse tree from it.
+   * Parse errors are output to {@link ToolIO#out}. Future work might switch
+   * this to the {@link Errors} class for uniformity & improved testability.
+   *
+   * @param input The input to parse, as a string.
+   * @return A parse tree, or null if syntax error.
+   */
+  public static SyntaxTreeNode processSyntax(String input) {
+    byte[] inputBytes = input.getBytes(StandardCharsets.UTF_8);
+    InputStream inputStream = new ByteArrayInputStream(inputBytes);
+    return processSyntax(inputStream);
+  }
+
+  /**
+   * Checks the syntax of the given input and builds a parse tree from it.
+   *
+   * @param input The input to parse, as a stream.
+   * @return A syntax parse tree, or null if parsing error.
+   */
+  public static SyntaxTreeNode processSyntax(InputStream input) {
+    TLAplusParser parser = new TLAplusParser(input, StandardCharsets.UTF_8.name());
+    return parser.parse() ? parser.ParseTree : null;
+  }
+
+  /**
+   * Performs semantic checking & reference resolution on a parse tree.
+   *
+   * @param parseTree The parse tree to check.
+   * @param log A log to record warnings & errors.
+   * @return A semantic parse tree, or null if semantic error.
+   */
+  public static ModuleNode processSemantics(SyntaxTreeNode parseTree, Errors log) {
+    // These two lines are annoying incantations to set global static state
+    // that will hopefully be made unnecessary in the future.
+    Context.reInit();
+    SemanticNode.setError(log);
+    // The null parameter here is the {@link ExternalModuleTable}, which will
+    // need to be resolved & populated if this API is to one day support TLA+
+    // modules that EXTEND or INSTANCE other TLA+ modules.
+    Generator gen = new Generator(null, log);
+    try {
+      ModuleNode semanticRoot = gen.generate(parseTree);
+      return log.isSuccess() ? semanticRoot : null;
+    } catch (AbortException e) {
+      Assert.fail(e.toString() + log.toString());
```

Comment:
Why does this catch the specific `AbortException` to replace it with the generic `TLCRuntimeException`?

Reply:
Temporary. The error handling of this API is not yet figured out. Probably it will eventually be much more elaborate as https://github.com/tlaplus/rfcs/issues/15 is worked on.

Reply:
Please use TODO markers for incomplete work.